data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/buffer/KafkaBuffer.java

```diff
@@ -0,0 +1,91 @@
+package org.opensearch.dataprepper.plugins.kafka.buffer;
+
+import org.opensearch.dataprepper.metrics.PluginMetrics;
+import org.opensearch.dataprepper.model.CheckpointState;
+import org.opensearch.dataprepper.model.acknowledgements.AcknowledgementSetManager;
+import org.opensearch.dataprepper.model.annotations.DataPrepperPlugin;
+import org.opensearch.dataprepper.model.annotations.DataPrepperPluginConstructor;
+import org.opensearch.dataprepper.model.buffer.AbstractBuffer;
+import org.opensearch.dataprepper.model.buffer.Buffer;
+import org.opensearch.dataprepper.model.configuration.PluginSetting;
+import org.opensearch.dataprepper.model.plugin.PluginFactory;
+import org.opensearch.dataprepper.model.record.Record;
+import org.opensearch.dataprepper.plugins.buffer.blockingbuffer.BlockingBuffer;
+import org.opensearch.dataprepper.plugins.kafka.configuration.KafkaBufferConfig;
+import org.opensearch.dataprepper.plugins.kafka.consumer.KafkaCustomConsumer;
+import org.opensearch.dataprepper.plugins.kafka.consumer.KafkaCustomConsumerFactory;
+import org.opensearch.dataprepper.plugins.kafka.producer.KafkaCustomProducer;
+import org.opensearch.dataprepper.plugins.kafka.producer.KafkaCustomProducerFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeoutException;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+@DataPrepperPlugin(name = "kafka_buffer", pluginType = Buffer.class, pluginConfigurationType = KafkaBufferConfig.class)
+public class KafkaBuffer<T extends Record<?>> extends AbstractBuffer<T> {
+
+    private static final Logger LOG = LoggerFactory.getLogger(KafkaBuffer.class);
+    private final KafkaCustomProducer producer;
+    private final AbstractBuffer innerBuffer;
+    private final ExecutorService executorService;
+
+    @DataPrepperPluginConstructor
+    public KafkaBuffer(final PluginSetting pluginSetting, final KafkaBufferConfig kafkaBufferConfig, final PluginFactory pluginFactory,
+                       final AcknowledgementSetManager acknowledgementSetManager, final PluginMetrics pluginMetrics){
+        super(pluginSetting);
+        final KafkaCustomProducerFactory kafkaCustomProducerFactory = new KafkaCustomProducerFactory();
+        producer = kafkaCustomProducerFactory.createProducer(kafkaBufferConfig, pluginFactory, pluginSetting,  null, null);
+        final KafkaCustomConsumerFactory kafkaCustomConsumerFactory = new KafkaCustomConsumerFactory();
+        innerBuffer = new BlockingBuffer<>(1000000, 250000, pluginSetting.getPipelineName());
```

Comment:
We might be able to get away without an inner buffer since the consumer will buffer on its own. I think we are fine to leave this for now and we can come back to it.

Let's pull the buffer and batch size out into constants. If we do go the inner buffer route, they should be configurable similar to the BlockingBuffer

Reply:
I was looking into this, is the consumer buffer `BufferAccumulator<Record<Event>> bufferAccumulator` defined in KafkaCustomConsumer? That class also has `private final Buffer<Record<Event>> buffer` but I'm assuming that is typically where the pipeline buffer would go when using Kafka source?